versioned/storage/cleanup/src/main/java/org/projectnessie/versioned/storage/cleanup/CutHistoryImpl.java

```diff
@@ -0,0 +1,150 @@
+    // before or after their children.
+    try (CloseableIterator<Obj> scan = persist.scanAllObjects(Collections.singleton(COMMIT))) {
+      while (scan.hasNext()) {
+        scanRateLimiter.acquire();
+
+        CommitObj commit = (CommitObj) scan.next();
+        numObjects++;
+
+        var tail = commit.tail();
+        var cutPointIdx = tail.indexOf(cutPoint);
+        // Commits referencing the cut point as their last parent do not need to be rewritten
+        if (cutPointIdx >= 0 && cutPointIdx < tail.size() - 1) {
+          result.addAffectedCommitId(commit.id());
+        }
+      }
+
+      result.numScannedObjs(numObjects);
+      return result.build();
+    }
+  }
+
+  @Override
+  public CutHistoryResult cutHistory(CutHistoryScanResult scanResult) {
+    var result = CutHistoryResult.builder().input(scanResult);
+    var commitLogic = commitLogic(context.persist());
+    var cutPoint = context.cutPoint();
+
+    result.wasHistoryCut(false);
+    var failed = false;
+    for (ObjId id : scanResult.affectedCommitIds()) {
+      try {
+        var commitObj = commitLogic.fetchCommit(id);
+        checkState(commitObj != null, "Commit not found: %s", id);
+
+        if (commitObj.tail().size() <= 1) {
+          continue;
+        }
+
+        var tail = commitObj.tail();
+        var cutPointIdx = tail.indexOf(cutPoint);
+
+        if (cutPointIdx < 0 || cutPointIdx == tail.size() - 1) {
+          // may happen on re-invocation
+          continue;
+        }
+
+        var updatedCommitObj =
+            CommitObj.commitBuilder()
+                .from(commitObj)
+                .tail(commitObj.tail().subList(0, cutPointIdx + 1))
+                .build();
+
+        if (!context.dryRun()) {
+          context.persist().upsertObj(updatedCommitObj);
+          result.addRewrittenCommitId(id);
+        }
+      } catch (Exception e) {
+        result.putFailure(id, e);
+        failed = true;
+      }
+    }
+
+    // Do not attempt cutting history if the affected commit tails could not be rewritten
+    if (failed) {
+      return result.build();
+    }
+
+    try {
+      var commitObj = commitLogic.fetchCommit(cutPoint);
+      checkState(commitObj != null, "Commit not found: %s", cutPoint);
+
+      if (!EMPTY_OBJ_ID.equals(commitObj.directParent())
+          || !commitObj.secondaryParents().isEmpty()) {
+        var msg =
+            String.format(
+                "%s [updated to remove parents on %s]",
+                commitObj.message(), context.persist().config().clock().instant());
+        var updatedCommitObj =
+            CommitObj.commitBuilder()
+                .from(commitObj)
+                .message(msg)
+                .secondaryParents(List.of())
+                .tail(List.of(EMPTY_OBJ_ID))
+                .build();
+
+        if (!context.dryRun()) {
+          context.persist().upsertObj(updatedCommitObj);
+          result.addRewrittenCommitId(cutPoint);
+          result.wasHistoryCut(true);
```

Comment:
Maybe always add both to the result, even in dry-run

Reply:
These IDs can be obtained from the related `CutHistoryScanResult`.